File: src/go/fab/internal/archive/archive.go
package archive

import (
	"bufio"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/wvrdz/fab-kit/src/go/fab/internal/change"
	"github.com/wvrdz/fab-kit/src/go/fab/internal/resolve"
)

// ArchiveResult holds the YAML output for archive operations.
type ArchiveResult struct {
	Action  string
	Name    string
	Clean   string
	Move    string
	Index   string
	Pointer string
}

// RestoreResult holds the YAML output for restore operations.
type RestoreResult struct {
	Action  string
	Name    string
	Move    string
	Index   string
	Pointer string
}

// parseDateBucket extracts yyyy and mm from a YYMMDD-prefixed folder name.
func parseDateBucket(name string) (string, string, error) {
	if len(name) < 6 {
		return "", "", fmt.Errorf("invalid folder name '%s': expected YYMMDD prefix", name)
	}
	for _, c := range name[:6] {
		if c < '0' || c > '9' {
			return "", "", fmt.Errorf("invalid folder name '%s': expected YYMMDD prefix", name)
		}
	}
	yy := name[0:2]
	mm := name[2:4]
	return "20" + yy, mm, nil
}

// Archive moves a change to the archive directory.
func Archive(fabRoot, changeArg, description string) (*ArchiveResult, error) {
	if changeArg == "" {
		return nil, fmt.Errorf("<change> argument is required for archive")
	}
	if description == "" {
		return nil, fmt.Errorf("--description is required for archive")
	}

	folder, err := resolve.ToFolder(fabRoot, changeArg)
	if err != nil {
		return nil, err
	}

	changesDir := filepath.Join(fabRoot, "changes")
	archiveDir := filepath.Join(changesDir, "archive")
	changeDir := filepath.Join(changesDir, folder)

	// 1. Clean: delete .pr-done if present
	cleanStatus := "not_present"
	prDonePath := filepath.Join(changeDir, ".pr-done")
	if _, err := os.Stat(prDonePath); err == nil {
		os.Remove(prDonePath)
		cleanStatus = "removed"
	}

	// 2. Move to archive/yyyy/mm/
	bucketYear, bucketMonth, err := parseDateBucket(folder)
	if err != nil {
		return nil, err
	}
	destDir := filepath.Join(archiveDir, bucketYear, bucketMonth)
	os.MkdirAll(destDir, 0755)
	destPath := filepath.Join(destDir, folder)
	if _, err := os.Stat(destPath); err == nil {
		return nil, fmt.Errorf("Archive destination already exists: %s", destPath)
	}
	if err := os.Rename(changeDir, destPath); err != nil {
		return nil, fmt.Errorf("move to archive: %w", err)
	}

	// 3. Update index
	indexFile := filepath.Join(archiveDir, "index.md")
	indexStatus := updateIndex(indexFile, folder, description)

	// Backfill unindexed
	backfillIndex(archiveDir, indexFile)

	// 4. Clear pointer if active
	pointerStatus := "skipped"
	activeFolder, err := resolve.ToFolder(fabRoot, "")
	if err == nil && activeFolder == folder {
		change.SwitchBlank(fabRoot)
		pointerStatus = "cleared"
	}

	return &ArchiveResult{
		Action:  "archive",
		Name:    folder,
		Clean:   cleanStatus,
		Move:    "moved",
		Index:   indexStatus,
		Pointer: pointerStatus,
	}, nil
}

// Restore moves a change from the archive back to active.
func Restore(fabRoot, changeArg string, doSwitch bool) (*RestoreResult, error) {
	if changeArg == "" {
		return nil, fmt.Errorf("<change> argument is required for restore")
	}

	folder, resolvedDir, err := resolveArchive(fabRoot, changeArg)
	if err != nil {
		return nil, err
	}

	changesDir := filepath.Join(fabRoot, "changes")
	archiveDir := filepath.Join(changesDir, "archive")

	// 1. Move from archive (resolveArchive returns full dir path)
	moveStatus := "restored"
	destPath := filepath.Join(changesDir, folder)
	if _, err := os.Stat(destPath); err == nil {
		moveStatus = "already_in_changes"
	} else {
		if err := os.Rename(resolvedDir, destPath); err != nil {
			return nil, fmt.Errorf("restore: %w", err)
		}
	}

	// 2. Remove from index
	indexFile := filepath.Join(archiveDir, "index.md")
	indexStatus := removeFromIndex(indexFile, folder)

	// 3. Optionally switch
	pointerStatus := "skipped"
	if doSwitch {
		_, err := change.Switch(fabRoot, folder)
		if err == nil {
			pointerStatus = "switched"
		}
	}

	return &RestoreResult{
		Action:  "restore",
		Name:    folder,
		Move:    moveStatus,
		Index:   indexStatus,
		Pointer: pointerStatus,
	}, nil
}

// List returns archived change folder names from both flat and nested entries.
func List(fabRoot string) ([]string, error) {
	archiveDir := filepath.Join(fabRoot, "changes", "archive")
	if _, err := os.Stat(archiveDir); os.IsNotExist(err) {
		return nil, nil
	}

	topLevel, err := os.ReadDir(archiveDir)
	if err != nil {
		return nil, err
	}

	var results []string

	// Flat entries: archive/{name}/ (skip year directories)
	for _, e := range topLevel {
		if e.IsDir() && !isYearDir(e.Name()) {
			results = append(results, e.Name())
		}
	}

	// Nested entries: archive/yyyy/mm/{name}/
	for _, yearEntry := range topLevel {
		if !yearEntry.IsDir() || !isYearDir(yearEntry.Name()) {
			continue
		}
		yearDir := filepath.Join(archiveDir, yearEntry.Name())
		monthEntries, _ := os.ReadDir(yearDir)
		for _, monthEntry := range monthEntries {
			if !monthEntry.IsDir() {
				continue
			}
			monthDir := filepath.Join(yearDir, monthEntry.Name())
			changeEntries, _ := os.ReadDir(monthDir)
			for _, ce := range changeEntries {
				if ce.IsDir() {
					results = append(results, ce.Name())
				}
			}
		}
	}

	return results, nil
}

// FormatArchiveYAML formats an ArchiveResult.
func FormatArchiveYAML(r *ArchiveResult) string {
	return fmt.Sprintf("action: %s\nname: %s\nclean: %s\nmove: %s\nindex: %s\npointer: %s",
		r.Action, r.Name, r.Clean, r.Move, r.Index, r.Pointer)
}

// FormatRestoreYAML formats a RestoreResult.
func FormatRestoreYAML(r *RestoreResult) string {
	return fmt.Sprintf("action: %s\nname: %s\nmove: %s\nindex: %s\npointer: %s",
		r.Action, r.Name, r.Move, r.Index, r.Pointer)
}

// resolveArchive returns (folderName, fullDirPath, error).
// Scans both flat entries (archive/{name}/) and nested entries (archive/yyyy/mm/{name}/).
func resolveArchive(fabRoot, override string) (string, string, error) {
	if override == "" {
		return "", "", fmt.Errorf("<change> argument is required for restore")
	}

	archiveDir := filepath.Join(fabRoot, "changes", "archive")
	if _, err := os.Stat(archiveDir); os.IsNotExist(err) {
		return "", "", fmt.Errorf("No archive folder found.")
	}

	type entry struct {
		name string
		dir  string
	}
	var entries []entry

	// Flat entries: archive/{name}/ (skip 4-digit year directories)
	topLevel, _ := os.ReadDir(archiveDir)
	for _, e := range topLevel {
		if !e.IsDir() {
			continue
		}
		name := e.Name()
		if isYearDir(name) {
			continue
		}
		entries = append(entries, entry{name, filepath.Join(archiveDir, name)})
	}

	// Nested entries: archive/yyyy/mm/{name}/
	for _, yearEntry := range topLevel {
		if !yearEntry.IsDir() || !isYearDir(yearEntry.Name()) {
			continue
		}
		yearDir := filepath.Join(archiveDir, yearEntry.Name())
		monthEntries, _ := os.ReadDir(yearDir)
		for _, monthEntry := range monthEntries {
			if !monthEntry.IsDir() {
				continue
			}
			monthDir := filepath.Join(yearDir, monthEntry.Name())
			changeEntries, _ := os.ReadDir(monthDir)
			for _, ce := range changeEntries {
				if ce.IsDir() {
					entries = append(entries, entry{ce.Name(), filepath.Join(monthDir, ce.Name())})
				}
			}
		}
	}

	if len(entries) == 0 {
		return "", "", fmt.Errorf("No archived changes found.")
	}

	overrideLower := strings.ToLower(override)

	// Exact match
	for _, e := range entries {
		if strings.ToLower(e.name) == overrideLower {
			return e.name, e.dir, nil
		}
	}

	// Substring match
	var partials []entry
	for _, e := range entries {
		if strings.Contains(strings.ToLower(e.name), overrideLower) {
			partials = append(partials, e)
		}
	}

	if len(partials) == 1 {
		return partials[0].name, partials[0].dir, nil
	}
	if len(partials) > 1 {
		names := make([]string, len(partials))
		for i, p := range partials {
			names[i] = p.name
		}
		return "", "", fmt.Errorf("Multiple archives match \"%s\": %s.", override, strings.Join(names, ", "))
	}

	return "", "", fmt.Errorf("No archive matches \"%s\".", override)
}

// isYearDir returns true if the name is a 4-digit year directory.
func isYearDir(name string) bool {
	if len(name) != 4 {
		return false
	}
	for _, c := range name {
		if c < '0' || c > '9' {
			return false
		}
	}
	return true
}

func updateIndex(indexFile, folder, description string) string {
	indexStatus := "updated"
	if _, err := os.Stat(indexFile); os.IsNotExist(err) {
		os.WriteFile(indexFile, []byte("# Archive Index\n\n"), 0644)
		indexStatus = "created"
	}

	// Normalize description
	description = strings.Map(func(r rune) rune {
		if r == '\n' || r == '\r' || r == '\t' {
			return ' '
		}
		return r
	}, description)
	description = strings.TrimSpace(description)

	newEntry := fmt.Sprintf("- **%s** — %s", folder, description)

	data, _ := os.ReadFile(indexFile)
	lines := strings.Split(string(data), "\n")

	var result []string
	if len(lines) >= 2 {
		result = append(result, lines[0], lines[1])
	} else if len(lines) == 1 {
		result = append(result, lines[0], "")
	} else {
		result = append(result, "# Archive Index", "")
	}
	result = append(result, newEntry)
	if len(lines) > 2 {
		result = append(result, lines[2:]...)
	}

	os.WriteFile(indexFile, []byte(strings.Join(result, "\n")), 0644)
	return indexStatus
}

func backfillIndex(archiveDir, indexFile string) {
	indexData, _ := os.ReadFile(indexFile)
	indexContent := string(indexData)

	f, err := os.OpenFile(indexFile, os.O_APPEND|os.O_WRONLY, 0644)
	if err != nil {
		return
	}
	defer f.Close()

	backfillEntry := func(name string) {
		marker := fmt.Sprintf("**%s**", name)
		if !strings.Contains(indexContent, marker) {
			fmt.Fprintf(f, "- **%s** — (no description — pre-index archive)\n", name)
		}
	}

	topLevel, _ := os.ReadDir(archiveDir)

	// Flat entries (pre-migration)
	for _, e := range topLevel {
		if e.IsDir() && !isYearDir(e.Name()) {
			backfillEntry(e.Name())
		}
	}

	// Nested entries: archive/yyyy/mm/{name}/
	for _, yearEntry := range topLevel {
		if !yearEntry.IsDir() || !isYearDir(yearEntry.Name()) {
			continue
		}
		yearDir := filepath.Join(archiveDir, yearEntry.Name())
		monthEntries, _ := os.ReadDir(yearDir)
		for _, monthEntry := range monthEntries {
			if !monthEntry.IsDir() {
				continue
			}
			monthDir := filepath.Join(yearDir, monthEntry.Name())
			changeEntries, _ := os.ReadDir(monthDir)
			for _, ce := range changeEntries {
				if ce.IsDir() {
					backfillEntry(ce.Name())
				}
			}
		}
	}
}

func removeFromIndex(indexFile, folder string) string {
	if _, err := os.Stat(indexFile); os.IsNotExist(err) {
		return "not_found"
	}

	marker := fmt.Sprintf("**%s**", folder)

	f, err := os.Open(indexFile)
	if err != nil {
		return "not_found"
	}

	var found bool
	var lines []string
	scanner := bufio.NewScanner(f)
	for scanner.Scan() {
		line := scanner.Text()
		if strings.Contains(line, marker) {
			found = true
			continue
		}
		lines = append(lines, line)
	}
	f.Close()

	if !found {
		return "not_found"
	}

	os.WriteFile(indexFile, []byte(strings.Join(lines, "\n")), 0644)
	return "removed"
}

File: src/fab-rust/src/archive.rs
use anyhow::{bail, Result};
use std::fs;
use std::io::{BufRead, BufReader, Write};
use std::path::{Path, PathBuf};

use crate::change;
use crate::resolve;

/// Archive result.
pub struct ArchiveResult {
    pub action: String,
    pub name: String,
    pub clean: String,
    pub mov: String,
    pub index: String,
    pub pointer: String,
}

/// Restore result.
pub struct RestoreResult {
    pub action: String,
    pub name: String,
    pub mov: String,
    pub index: String,
    pub pointer: String,
}

/// Moves a change to the archive directory.
pub fn archive(fab_root: &str, change_arg: &str, description: &str) -> Result<ArchiveResult> {
    if change_arg.is_empty() {
        bail!("<change> argument is required for archive");
    }
    if description.is_empty() {
        bail!("--description is required for archive");
    }

    let folder = resolve::to_folder(fab_root, change_arg)?;

    let changes_dir = PathBuf::from(fab_root).join("changes");
    let archive_dir = changes_dir.join("archive");
    let change_dir = changes_dir.join(&folder);

    // 1. Clean: delete .pr-done if present
    let clean_status;
    let pr_done_path = change_dir.join(".pr-done");
    if pr_done_path.exists() {
        let _ = fs::remove_file(&pr_done_path);
        clean_status = "removed".to_string();
    } else {
        clean_status = "not_present".to_string();
    }

    // 2. Move to archive/yyyy/mm/
    let (bucket_year, bucket_month) = parse_date_bucket(&folder)?;
    let dest_dir = archive_dir.join(&bucket_year).join(&bucket_month);
    let _ = fs::create_dir_all(&dest_dir);
    let dest_path = dest_dir.join(&folder);
    if dest_path.exists() {
        bail!("Archive destination already exists: {}", dest_path.display());
    }
    fs::rename(&change_dir, &dest_path)?;

    // 3. Update index
    let index_file = archive_dir.join("index.md");
    let index_status = update_index(&index_file, &folder, description);

    // Backfill unindexed
    backfill_index(&archive_dir, &index_file);

    // 4. Clear pointer if active
    let mut pointer_status = "skipped".to_string();
    if let Ok(active_folder) = resolve::to_folder(fab_root, "") {
        if active_folder == folder {
            change::switch_blank(fab_root);
            pointer_status = "cleared".to_string();
        }
    }

    Ok(ArchiveResult {
        action: "archive".to_string(),
        name: folder,
        clean: clean_status,
        mov: "moved".to_string(),
        index: index_status,
        pointer: pointer_status,
    })
}

/// Restores a change from the archive.
pub fn restore(fab_root: &str, change_arg: &str, do_switch: bool) -> Result<RestoreResult> {
    if change_arg.is_empty() {
        bail!("<change> argument is required for restore");
    }

    let (folder, resolved_dir) = resolve_archive(fab_root, change_arg)?;

    let changes_dir = PathBuf::from(fab_root).join("changes");
    let archive_dir = changes_dir.join("archive");

    // 1. Move from archive
    let move_status;
    let dest_path = changes_dir.join(&folder);
    if dest_path.exists() {
        move_status = "already_in_changes".to_string();
    } else {
        fs::rename(&resolved_dir, &dest_path)?;
        move_status = "restored".to_string();
    }

    // 2. Remove from index
    let index_file = archive_dir.join("index.md");
    let index_status = remove_from_index(&index_file, &folder);

    // 3. Optionally switch
    let mut pointer_status = "skipped".to_string();
    if do_switch {
        if change::switch(fab_root, &folder).is_ok() {
            pointer_status = "switched".to_string();
        }
    }

    Ok(RestoreResult {
        action: "restore".to_string(),
        name: folder,
        mov: move_status,
        index: index_status,
        pointer: pointer_status,
    })
}

/// Lists archived change folder names from both flat and nested entries.
pub fn list(fab_root: &str) -> Result<Vec<String>> {
    let archive_dir = PathBuf::from(fab_root).join("changes").join("archive");
    if !archive_dir.exists() {
        return Ok(Vec::new());
    }

    let top_level = fs::read_dir(&archive_dir)?;
    let mut results = Vec::new();

    let mut top_entries: Vec<_> = top_level.flatten().collect();
    top_entries.sort_by_key(|e| e.file_name());

    // Flat entries: archive/{name}/ (skip year directories)
    for e in &top_entries {
        if e.file_type().map(|t| t.is_dir()).unwrap_or(false) {
            let name = e.file_name().to_string_lossy().to_string();
            if !is_year_dir(&name) {
                results.push(name);
            }
        }
    }

    // Nested entries: archive/yyyy/mm/{name}/
    for year_entry in &top_entries {
        if !year_entry.file_type().map(|t| t.is_dir()).unwrap_or(false) {
            continue;
        }
        let year_name = year_entry.file_name().to_string_lossy().to_string();
        if !is_year_dir(&year_name) {
            continue;
        }
        let year_dir = archive_dir.join(&year_name);
        if let Ok(month_entries) = fs::read_dir(&year_dir) {
            let mut months: Vec<_> = month_entries.flatten().collect();
            months.sort_by_key(|e| e.file_name());
            for month_entry in &months {
                if !month_entry.file_type().map(|t| t.is_dir()).unwrap_or(false) {
                    continue;
                }
                let month_dir = year_dir.join(month_entry.file_name());
                if let Ok(change_entries) = fs::read_dir(&month_dir) {
                    let mut changes: Vec<_> = change_entries.flatten().collect();
                    changes.sort_by_key(|e| e.file_name());
                    for ce in &changes {
                        if ce.file_type().map(|t| t.is_dir()).unwrap_or(false) {
                            results.push(ce.file_name().to_string_lossy().to_string());
                        }
                    }
                }
            }
        }
    }

    Ok(results)
}

/// Formats an ArchiveResult.
pub fn format_archive_yaml(r: &ArchiveResult) -> String {
    format!(
        "action: {}\nname: {}\nclean: {}\nmove: {}\nindex: {}\npointer: {}",
        r.action, r.name, r.clean, r.mov, r.index, r.pointer
    )
}

/// Formats a RestoreResult.
pub fn format_restore_yaml(r: &RestoreResult) -> String {
    format!(
        "action: {}\nname: {}\nmove: {}\nindex: {}\npointer: {}",
        r.action, r.name, r.mov, r.index, r.pointer
    )
}

fn parse_date_bucket(name: &str) -> Result<(String, String)> {
    if name.len() < 6 {
        bail!(
            "invalid folder name '{}': expected YYMMDD prefix",
            name
        );
    }
    let prefix = &name[..6];
    for c in prefix.chars() {
        if !c.is_ascii_digit() {
            bail!(
                "invalid folder name '{}': expected YYMMDD prefix",
                name
            );
        }
    }
    let yy = &prefix[0..2];
    let mm = &prefix[2..4];
    Ok((format!("20{}", yy), mm.to_string()))
}

fn resolve_archive(fab_root: &str, override_arg: &str) -> Result<(String, PathBuf)> {
    if override_arg.is_empty() {
        bail!("<change> argument is required for restore");
    }

    let archive_dir = PathBuf::from(fab_root).join("changes").join("archive");
    if !archive_dir.exists() {
        bail!("No archive folder found.");
    }

    struct Entry {
        name: String,
        dir: PathBuf,
    }
    let mut entries = Vec::new();

    // Flat entries
    if let Ok(top_level) = fs::read_dir(&archive_dir) {
        for e in top_level.flatten() {
            if !e.file_type().map(|t| t.is_dir()).unwrap_or(false) {
                continue;
            }
            let name = e.file_name().to_string_lossy().to_string();
            if is_year_dir(&name) {
                continue;
            }
            entries.push(Entry {
                name,
                dir: e.path(),
            });
        }
    }

    // Nested entries
    if let Ok(top_level) = fs::read_dir(&archive_dir) {
        for year_entry in top_level.flatten() {
            if !year_entry.file_type().map(|t| t.is_dir()).unwrap_or(false) {
                continue;
            }
            let year_name = year_entry.file_name().to_string_lossy().to_string();
            if !is_year_dir(&year_name) {
                continue;
            }
            let year_dir = archive_dir.join(&year_name);
            if let Ok(month_entries) = fs::read_dir(&year_dir) {
                for month_entry in month_entries.flatten() {
                    if !month_entry.file_type().map(|t| t.is_dir()).unwrap_or(false) {
                        continue;
                    }
                    let month_dir = year_dir.join(month_entry.file_name());
                    if let Ok(change_entries) = fs::read_dir(&month_dir) {
                        for ce in change_entries.flatten() {
                            if ce.file_type().map(|t| t.is_dir()).unwrap_or(false) {
                                entries.push(Entry {
                                    name: ce.file_name().to_string_lossy().to_string(),
                                    dir: ce.path(),
                                });
                            }
                        }
                    }
                }
            }
        }
    }

    if entries.is_empty() {
        bail!("No archived changes found.");
    }

    let override_lower = override_arg.to_lowercase();

    // Exact match
    for e in &entries {
        if e.name.to_lowercase() == override_lower {
            return Ok((e.name.clone(), e.dir.clone()));
        }
    }

    // Substring match
    let partials: Vec<&Entry> = entries
        .iter()
        .filter(|e| e.name.to_lowercase().contains(&override_lower))
        .collect();

    if partials.len() == 1 {
        return Ok((partials[0].name.clone(), partials[0].dir.clone()));
    }
    if partials.len() > 1 {
        let names: Vec<&str> = partials.iter().map(|e| e.name.as_str()).collect();
        bail!(
            "Multiple archives match \"{}\": {}.",
            override_arg,
            names.join(", ")
        );
    }

    bail!("No archive matches \"{}\".", override_arg);
}

fn is_year_dir(name: &str) -> bool {
    if name.len() != 4 {
        return false;
    }
    name.chars().all(|c| c.is_ascii_digit())
}

fn update_index(index_file: &Path, folder: &str, description: &str) -> String {
    let mut index_status = "updated".to_string();
    if !index_file.exists() {
        let _ = fs::write(index_file, "# Archive Index\n\n");
        index_status = "created".to_string();
    }

    // Normalize description
    let description: String = description
        .chars()
        .map(|c| if c == '\n' || c == '\r' || c == '\t' { ' ' } else { c })
        .collect();
    let description = description.trim();

    let new_entry = format!("- **{}** \u{2014} {}", folder, description);

    let data = fs::read_to_string(index_file).unwrap_or_default();
    let lines: Vec<&str> = data.split('\n').collect();

    let mut result = Vec::new();
    if lines.len() >= 2 {
        result.push(lines[0].to_string());
        result.push(lines[1].to_string());
    } else if lines.len() == 1 {
        result.push(lines[0].to_string());
        result.push(String::new());
    } else {
        result.push("# Archive Index".to_string());
        result.push(String::new());
    }
    result.push(new_entry);
    if lines.len() > 2 {
        for line in &lines[2..] {
            result.push(line.to_string());
        }
    }

    let _ = fs::write(index_file, result.join("\n"));
    index_status
}

fn backfill_index(archive_dir: &Path, index_file: &Path) {
    let index_content = fs::read_to_string(index_file).unwrap_or_default();

    let mut f = match fs::OpenOptions::new().append(true).open(index_file) {
        Ok(f) => f,
        Err(_) => return,
    };

    let backfill_entry = |f: &mut fs::File, name: &str| {
        let marker = format!("**{}**", name);
        if !index_content.contains(&marker) {
            let _ = writeln!(
                f,
                "- **{}** \u{2014} (no description \u{2014} pre-index archive)",
                name
            );
        }
    };

    // Flat entries
    if let Ok(top_level) = fs::read_dir(archive_dir) {
        for e in top_level.flatten() {
            if e.file_type().map(|t| t.is_dir()).unwrap_or(false) {
                let name = e.file_name().to_string_lossy().to_string();
                if !is_year_dir(&name) {
                    backfill_entry(&mut f, &name);
                }
            }
        }
    }

    // Nested entries
    if let Ok(top_level) = fs::read_dir(archive_dir) {
        for year_entry in top_level.flatten() {
            if !year_entry.file_type().map(|t| t.is_dir()).unwrap_or(false) {
                continue;
            }
            let year_name = year_entry.file_name().to_string_lossy().to_string();
            if !is_year_dir(&year_name) {
                continue;
            }
            let year_dir = archive_dir.join(&year_name);
            if let Ok(month_entries) = fs::read_dir(&year_dir) {
                for month_entry in month_entries.flatten() {
                    if !month_entry.file_type().map(|t| t.is_dir()).unwrap_or(false) {
                        continue;
                    }
                    let month_dir = year_dir.join(month_entry.file_name());
                    if let Ok(change_entries) = fs::read_dir(&month_dir) {
                        for ce in change_entries.flatten() {
                            if ce.file_type().map(|t| t.is_dir()).unwrap_or(false) {
                                backfill_entry(
                                    &mut f,
                                    &ce.file_name().to_string_lossy(),
                                );
                            }
                        }
                    }
                }
            }
        }
    }
}

fn remove_from_index(index_file: &Path, folder: &str) -> String {
    if !index_file.exists() {
        return "not_found".to_string();
    }

    let marker = format!("**{}**", folder);

    let f = match fs::File::open(index_file) {
        Ok(f) => f,
        Err(_) => return "not_found".to_string(),
    };

    let mut found = false;
    let mut lines = Vec::new();
    let reader = BufReader::new(f);
    for line in reader.lines().flatten() {
        if line.contains(&marker) {
            found = true;
            continue;
        }
        lines.push(line);
    }

    if !found {
        return "not_found".to_string();
    }

    let _ = fs::write(index_file, lines.join("\n"));
    "removed".to_string()
}
